From 9999999999999999999999999999999999999999 Mon Sep 17 00:00:00 2001
From: Sarthak <sinhasarthak56@gmail.com>
Date: Wed, 15 Oct 2026 10:45:00 +0530
Subject: [PATCH 9/9] ImsService: Force update IMS config on carrier config
 loaded and MMTEL ready

- In Android 17 / AOSP, CarrierConfigManager.getDefaultConfig() has
  carrier_volte_available_bool as false by default.
- ImsConfigEventReceiver was checking this default config before carrier
  config loaded and aborting without updating IMS config.
- Furthermore, handleCarrierConfigChanged aborted on hasIccCard() check
  and never called updateImsServiceConfig after carrier config finished loading.
- On SIM_STATE_CHANGED, dynamic_ims_switch and defaultSupportVolte checks
  prevented updating IMS config when SIM loaded.
- ACTION_CARRIER_CONFIG_CHANGED in AOSP may only include EXTRA_SLOT_INDEX
  instead of "phone" extra.
- Trigger onReceive with ACTION_MTK_MMTEL_READY immediately upon
  ImsConfigController initialization and bypass early aborts.

Signed-off-by: Sarthak <sinhasarthak56@gmail.com>
---
 .../ims/config/internal/ImsConfigController.smali  | 18 ++++++
 .../config/internal/ImsConfigEventReceiver.smali   | 66 +++++-----------------
 2 files changed, 33 insertions(+), 51 deletions(-)

--- a/smali/com/mediatek/ims/config/internal/ImsConfigController.smali
+++ b/smali/com/mediatek/ims/config/internal/ImsConfigController.smali
@@ -528,6 +528,24 @@
 
     invoke-virtual {v4, v5, v3, v7}, Landroid/content/Context;->registerReceiver(Landroid/content/BroadcastReceiver;Landroid/content/IntentFilter;I)Landroid/content/Intent;
 
+    new-instance v3, Landroid/content/Intent;
+
+    const-string v4, "com.mediatek.ims.MTK_MMTEL_READY"
+
+    invoke-direct {v3, v4}, Landroid/content/Intent;-><init>(Ljava/lang/String;)V
+
+    const-string v4, "android:phone_id"
+
+    iget v5, p0, Lcom/mediatek/ims/config/internal/ImsConfigController;->mPhoneId:I
+
+    invoke-virtual {v3, v4, v5}, Landroid/content/Intent;->putExtra(Ljava/lang/String;I)Landroid/content/Intent;
+
+    iget-object v4, p0, Lcom/mediatek/ims/config/internal/ImsConfigController;->mReceiver:Landroid/content/BroadcastReceiver;
+
+    iget-object v5, p0, Lcom/mediatek/ims/config/internal/ImsConfigController;->mContext:Landroid/content/Context;
+
+    invoke-virtual {v4, v5, v3}, Landroid/content/BroadcastReceiver;->onReceive(Landroid/content/Context;Landroid/content/Intent;)V
+
     .line 158
     iget-object v4, p0, Lcom/mediatek/ims/config/internal/ImsConfigController;->mRilAdapter:Lcom/mediatek/ims/ril/ImsCommandsInterface;
 
--- a/smali/com/mediatek/ims/config/internal/ImsConfigEventReceiver.smali
+++ b/smali/com/mediatek/ims/config/internal/ImsConfigEventReceiver.smali
@@ -245,6 +245,20 @@
 
     .line 267
     .local v0, "phoneId":I
+    if-ne v0, v1, :cond_check_slot
+
+    const-string v0, "android.telephony.extra.SLOT_INDEX"
+
+    invoke-virtual {p2, v0, v1}, Landroid/content/Intent;->getIntExtra(Ljava/lang/String;I)I
+
+    move-result v0
+
+    :cond_check_slot
+    if-ne v0, v1, :cond_check_match
+
+    iget v0, p0, Lcom/mediatek/ims/config/internal/ImsConfigEventReceiver;->mPhoneId:I
+
+    :cond_check_match
     iget v1, p0, Lcom/mediatek/ims/config/internal/ImsConfigEventReceiver;->mPhoneId:I
 
     if-eq v0, v1, :cond_c
@@ -314,24 +328,7 @@
 
     .line 279
     .local v2, "tm":Landroid/telephony/TelephonyManager;
-    if-eqz v2, :cond_51
-
-    .line 280
-    invoke-virtual {v2, v0}, Landroid/telephony/TelephonyManager;->hasIccCard(I)Z
-
-    move-result v3
-
-    if-nez v3, :cond_51
-
-    .line 284
-    iget-object v3, p0, Lcom/mediatek/ims/config/internal/ImsConfigEventReceiver;->mLogTag:Ljava/lang/String;
-
-    const-string v4, "No need to reload config storage"
-
-    invoke-static {v3, v4}, Landroid/telephony/Rlog;->e(Ljava/lang/String;Ljava/lang/String;)I
-
-    .line 285
-    return-void
+    goto :cond_51
 
     .line 289
     :cond_51
@@ -525,6 +522,8 @@
     :goto_fe
     invoke-direct {p0, p1, v0}, Lcom/mediatek/ims/config/internal/ImsConfigEventReceiver;->setVoiceDomainSetting(Landroid/content/Context;I)V
 
+    invoke-direct {p0, p1, v0}, Lcom/mediatek/ims/config/internal/ImsConfigEventReceiver;->updateImsServiceConfig(Landroid/content/Context;I)V
+
     .line 315
     return-void
 .end method
@@ -1080,6 +1079,8 @@
     invoke-direct {p0, v1}, Lcom/mediatek/ims/config/internal/ImsConfigEventReceiver;->updateFeatureSupportProperty(Landroid/os/Bundle;)V
 
     .line 190
+    iget v0, p0, Lcom/mediatek/ims/config/internal/ImsConfigEventReceiver;->mPhoneId:I
+
     invoke-direct {p0, p1, v0}, Lcom/mediatek/ims/config/internal/ImsConfigEventReceiver;->updateImsServiceConfig(Landroid/content/Context;I)V
 
     goto/16 :goto_1cd
@@ -1193,12 +1194,6 @@
 
     .line 171
     .local v3, "defaultSupportVolte":Z
-    if-nez v3, :cond_b9
-
-    iget-boolean v4, p0, Lcom/mediatek/ims/config/internal/ImsConfigEventReceiver;->mIsCarrierConfigLoaded:Z
-
-    if-eqz v4, :cond_b3
-
     goto :goto_b9
 
     .line 174
@@ -1314,12 +1309,6 @@
     invoke-static {v4, v5}, Landroid/telephony/Rlog;->d(Ljava/lang/String;Ljava/lang/String;)I
 
     .line 131
-    if-nez v3, :cond_122
-
-    iget-boolean v4, p0, Lcom/mediatek/ims/config/internal/ImsConfigEventReceiver;->mIsCarrierConfigLoaded:Z
-
-    if-eqz v4, :cond_11c
-
     goto :goto_122
 
     .line 134
@@ -1473,33 +1462,6 @@
 
     .line 102
     :cond_199
-    invoke-static {}, Landroid/telephony/CarrierConfigManager;->getDefaultConfig()Landroid/os/PersistableBundle;
-
-    move-result-object v2
-
-    invoke-virtual {v2, v6}, Landroid/os/PersistableBundle;->getBoolean(Ljava/lang/String;)Z
-
-    move-result v2
-
-    .line 104
-    .local v2, "defaultSupportVolte":Z
-    if-eqz v2, :cond_1cc
-
-    const-string v3, "persist.vendor.mtk_dynamic_ims_switch"
-
-    invoke-static {v3}, Landroid/os/SystemProperties;->get(Ljava/lang/String;)Ljava/lang/String;
-
-    move-result-object v3
-
-    const-string v4, "1"
-
-    invoke-virtual {v4, v3}, Ljava/lang/Object;->equals(Ljava/lang/Object;)Z
-
-    move-result v3
-
-    if-nez v3, :cond_1cc
-
-    .line 106
     iget-object v3, p0, Lcom/mediatek/ims/config/internal/ImsConfigEventReceiver;->mLogTag:Ljava/lang/String;
 
     new-instance v4, Ljava/lang/StringBuilder;
-- 
2.53.0
